src/embed/connection.rs

```diff
@@ -0,0 +1,256 @@
+    let txn = match txns.get(&txn_id) {
+        Some(v) => v,
+        None => {
+            return req
+                .response
+                .send(Err(format!("No transaction {}", txn_id)))
+                .await
+        }
+    };
+
+    req.response.send(func.call(txn, request).await).await;
+}
+
+async fn do_open<'a, 'b>(store: &'a dag::Store, txns: &'b TxnMap<'a>, req: Request) {
+    use OpenTransactionError::*;
+    let dag_write = store.write().await.map_err(DagWriteError).unwrap();
+    let write = db::Write::new_from_head("main", dag_write)
+        .await
+        .map_err(DBWriteError)
+        .unwrap();
+    let txn_id = TRANSACTION_COUNTER.fetch_add(1, Ordering::SeqCst);
+    txns.write().await.insert(txn_id, RwLock::new(write));
+    req.response
+        .send(Ok(SerJson::serialize_json(&OpenTransactionResponse {
+            transaction_id: txn_id,
+        })))
+        .await;
+}
+
+async fn do_commit(txns: &TxnMap<'_>, req: Request) {
+    let txn_id = match deserialize::<CommitTransactionRequest>(&req.data) {
+        Ok(v) => v.transaction_id,
+        Err(e) => return req.response.send(Err(e)).await,
+    };
+    let mut txns = txns.write().await;
+    let txn = match txns.remove(&txn_id) {
+        Some(v) => v,
+        None => {
+            return req
+                .response
+                .send(Err(format!("No such transaction {}", txn_id)))
+                .await;
+        }
+    };
+    let txn = txn.into_inner();
+    let response = txn
+        .commit(
+            "main",
+            "local-create-date",
+            "checksum",
+            42,
+            "foo",
+            b"bar",
+            None,
+        )
+        .await;
+    if let Err(e) = response {
+        req.response
+            .send(Err(format!("{:?}", CommitError::CommitError(e))))
+            .await;
+        return;
+    }
+    req.response
+        .send(Ok(SerJson::serialize_json(&CommitTransactionResponse {})))
+        .await;
+}
+
+async fn do_abort(txns: &TxnMap<'_>, req: Request) {
+    let request: CommitTransactionRequest = match DeJson::deserialize_json(&req.data) {
+        Ok(v) => v,
+        Err(e) => return req.response.send(Err(format!("InvalidJson({})", e))).await,
+    };
+    let txn_id = request.transaction_id;
+    if txns.write().await.remove(&txn_id).is_none() {
+        return req
+            .response
+            .send(Err(format!("No transaction {}", txn_id)))
+            .await;
+    };
+    req.response
+        .send(Ok(SerJson::serialize_json(&CommitTransactionResponse {})))
+        .await;
+}
+
+async fn do_has(txn: &RwLock<db::Write<'_>>, req: GetRequest) -> Result<String, String> {
+    Ok(SerJson::serialize_json(&GetResponse {
+        has: txn.read().await.has(req.key.as_bytes()),
+        value: None,
+    }))
+}
+
+async fn do_get(txn: &RwLock<db::Write<'_>>, req: GetRequest) -> Result<String, String> {
+    #[cfg(not(default))] // Not enabled in production.
```

Comment:
The intention is to never remove this; it's meant to provide a special behavior for tests, but can't be tagged `cfg(test)` because `wasm-pack test` doesn't set that.

I'd like to have a clearer way to indicate this, but I have found value so far in having small breaks in the black-box nature of the code when writing wasm tests.